src/toxcore/dht/server/mod.rs

```diff
@@ -219,32 +225,45 @@ impl Server {
         let ping_and_get_close_nodes = self.ping_and_get_close_nodes(Duration::from_secs(self.config.ping_interval));
         let send_nodes_req_random = self.send_nodes_req_random(Duration::from_secs(self.config.bad_node_timeout),
                                                                Duration::from_secs(self.config.nodes_req_interval));
-        let send_nodes_req_packets = {
-            let mut friends = self.friends.write();
-
-            let nodes_sender = friends.iter_mut()
-                .map(|friend| {
-                    friend.send_nodes_req_packets(self, Duration::from_secs(self.config.ping_interval),
-                                                  Duration::from_secs(self.config.nodes_req_interval),
-                                                  Duration::from_secs(self.config.bad_node_timeout))
-                });
+        let send_nodes_req_to_friends = self.send_nodes_req_to_friends();
 
-            let nodes_stream = stream::futures_unordered(nodes_sender).then(|_| Ok(()));
-
-            nodes_stream.for_each(|()| Ok(()))
-        };
+        let ping_sender = self.iterate_ping_list(Duration::from_secs(self.config.ping_iter_interval));
 
         let send_nat_ping_req = self.send_nat_ping_req(Duration::from_secs(self.config.nat_ping_req_interval));
 
-        let res = ping_bootstrap_nodes.join5(ping_and_get_close_nodes,
-                                             send_nodes_req_random,
-                                             send_nodes_req_packets,
-                                             send_nat_ping_req)
+        let res = future::join_all(vec![ping_bootstrap_nodes,
```

Comment:
You can use static array here instead of vector to not allocate the memory.

Reply:
I don't know how to